### ISSUE TITLE
Workspace

### ISSUE THREAD

**Issue by mjhenkes** [cerner/terra-application#85]
# Feature Request

## Description
Create a workspace component in terra-application v1.

The intent of the workspace area is to create a space separate from the page where users have in progress forms that relate to the content of the 'page'. This issue will track both the creation of the workspace component and the workspace tab component.

### Workspace

The tabs container for the workspace area.

* Contains one to many tabs
* Provides a dismiss button
* Provides a menu of options intended for:
  * size selection
  * open/close toggle
* Tabs must roll up in some way for small sizes

### Workspace Tab

The tab component that consumers fill out and provide to the secondary navigation layout.

* Always contains a non consumer controlled header
  * Workspace Tab title is kept in sync with workspace provided tab.
  * May have a spot for additional actions/menu (? confirm)
* Has a dedicated slot for a 'sticky' toolbar.
  * Any responsive behaviour for the toolbar must be managed by the consumer.
* Has notification banner
* Has a slot for consumer defined workspace content
* Has access to:
  * Notification banner context api
  * Loading Overlay context api
  * Status Overlay context api
  * ActiveMainPage Custom Hook 

#### Workspace API

Props 
* id 
* activeTabKey 
* onRequestActivate 
* ariaLabel – aria label associated with tab structure – TBD – need some UX feedback on whether this can be standardized or needs to be data dependent 
* activeSize – String mapping to a size option 
* sizeOptions – Array of key value pairs - ```{ key:<size>, text: <display>, isDisabled: <bool> }```
* onRequestSizeChange - (sizeKey) => {} 
* onRequestDismiss – optional, indicates a dismiss button should be present 
* children 
  *   WorkspaceTab
          * tabKey - Key unique within the tabs to used to selection display
          * label - Display or aria-label text for the tab
          * metaData - object returned wtihin the onRequestActivate
          * render - render function for the content within the tab panel
          * icon - Html node of the icon
          * isIconOnly - Whether or not icon is displayed only 

#### Consumer Implementation API

Exports From Individual Tab Component Module 
* Title for Tab – Intl key for the title to display for the tab's label prop 
* React Component 
  * Panel 
    * actions – Array 
      * key - String 
      * title - String 
      * isSelected – Boolean 
      * onAction - () => {} 
    * toolbar - Node 
    * children 

Available Custom Hooks 
* useActiveMainPage 
  * pageKey – this could also be pageCategoryMean 
  * pageDescription 
  * pageMetaData 
    * pageCategoryMean  

```jsx
const contextValue = {​​​​​​
  pageKey: 'page-1',
  pageLabel: 'Page 1',
  pageMetaData: {​​​​​​
    data: '1234',
  }​​​​​​,  
}​​​​​​;

return (
  <ActiveMainPageContext.Provider value={​​​​​​​contextValue}​​​​​​​>
    {​​​​​​​children}​​​​​​​
  </ActiveMainPageContext.Provider>
);
```

Available Contexts 
* TabContext 
  * isActive - Boolean 

Available Application Content Components 
* Navigation Prompt 
* Notification Banner 
* Application Loading Overlay 

### Layers

There are two scenarios that we are designing for: pure web terra-application and embedding inside of MPages.

The layers involved in the 2 scenarios will be:

Terra Application
* SecondaryNavigationLayout   
  * SecondaryNavigationLayoutWorkspace
    * Workspace
      * Consumer Tab Component
        * Panel

MPages
* MPageLayout
  * Dedicated div to contain workspace 
    * Workspace/EventEmitter Adapter 
      * Workspace
        * Consumer Tab Component
          * Panel

## Additional Context / Screenshots

Artistic rendering:
![Screen Shot 2020-09-10 at 2 36 39 PM](https://user-images.githubusercontent.com/1588747/92792337-0fcb6d00-f373-11ea-9192-6cb91a6021ed.png)

## @ Mentions

@dkasper-was-taken

**Comment by tbiethman** [comment]
Closed with #130
